### ISSUE TITLE
[WXGTK] UI label with msgctxt translation is overridden by non-translated stock labels

### ISSUE THREAD

**Issue by artur-so** [wxWidgets/wxWidgets#24366]
#### Bug description:

In an application that uses localization, buttons with standard IDs, e.g. `wxID_OK`, `wxID_CANCEL`, are considered stock items.
For stock buttons with stock labels the gtk stock labels are used. Those are not translated and they override the translated labels intended to be used by the application.
The code that produces the behavior is [here](https://github.com/wxWidgets/wxWidgets/blob/master/src/gtk/button.cpp#L258) .

There is a GTK mechanism for translating stock labels, but it is not used. See [gtk-docs](https://docs.gtk.org/gtk3/func.stock_set_translate_func.html).

On the other hand, GTK docs state that the stock functionality has been deprecated since 3.10.
In fact, https://github.com/wxWidgets/wxWidgets/commit/38c192acc5e000c91034dff3c563dadb3ebc642a disabled stock buttons. With this change the problem would be avoided.

However, for some reason in https://github.com/wxWidgets/wxWidgets/commit/29bd2c92747f7bd53c74610f0683cc3e91ffbc17 the GTK stock buttons are disabled only since GTK 4.

In my opinion, stock labels should be used only as fallback when client code sets no label.

#### Expected vs observed behaviour:

The application-provided translation should be used for labels of buttons with standard IDs and stock labels.
It is not reproducible with wxMSW and wxOSX.

#### Patch or snippet allowing to reproduce the problem:

Below is a diff to illustrate the scenario, but our case is more complex and involves mnemonics in xrc label.
```diff
Index: samples/internat/internat.cpp
===================================================================
--- samples/internat/internat.cpp
+++ samples/internat/internat.cpp
@@ -343,7 +343,7 @@
     wxSizer *sizer = new wxBoxSizer(wxHORIZONTAL);
     sizer->Add(new wxStaticText(this, wxID_ANY, _("First")),
                 wxSizerFlags().Border());
-    sizer->Add(new wxStaticText(this, wxID_ANY, _("Second")),
+    sizer->Add(new wxButton(this, wxID_CANCEL, _("&Cancel")),
                 wxSizerFlags().Border());
     SetSizer(sizer);
 }

```

### Platform and version information

- wxWidgets version  you use:  3.1.2
- wxWidgets port  you use: wxGTK
- OS  and its version: Ubuntu 22.04
  
  + GTK version:  3.24.33
  + Which GDK  backend is used: X11
  + Desktop environment : Gnome

**Comment by artur-so** [comment]
Updated the issue description so that is is easily reproducible with the head of the master branch.
This is not a regression, it is around for at least 13 years.
I guess it is a rather exotic case, but a real one.
At our side we fixed it by disabling the usage of stock items, that is by reverting https://github.com/wxWidgets/wxWidgets/commit/29bd2c92747f7bd53c74610f0683cc3e91ffbc17 .
The issue itself could be useful for the record, but I estimate it will hardly get fixed. It looks like it is at gtk side, although can be worked around.
The issue can be closed.

**Comment by artur-so** [comment]
Added screenshots,

**Comment by vadz** [comment]
I could be missing something here, but I thought that to use a translation defined in some context you had to specify this context when looking up the string, i.e. for this to work you'd have to use

```cpp
    help_menu->Append(wxID_ABOUT, wxGETTEXT_IN_CONTEXT("About context", "&About"));
```

in the sample code, don't you?

I also have no idea how can the referenced commit affect this, it seems to not change anything for the menu items at all (but only buttons)...

**Comment by artur-so** [comment]
@vadz thank you the comment! It is useful to get a better understanding of the context (pun intended 😄 ).
>to use a translation defined in some context you had to specify this context when looking up the string

Indeed, using `wxGETTEXT_IN_CONTEXT` results in the menu string being properly translated. I hadn't tried that.
In our application, there is more context that has impact here:
* we add the context solely as a translation hint; it is not (yet?) used to disambiguate different translations of the same string so the context translation mechanism is not made use of;
* the affected labels at our side are defined in xrc files and, after a cursory check, I don't see whether specifying a context is supported.

> it seems to not change anything for the menu items at all (but only buttons)...

You are right, that is confusing. It is caused by the fact that our scenario is different from the provided reproducible diff. At our side, buttons are affected (the reason for this issue). With 29bd2c92 reverted and 38c192ac in place the stock items are not used and the issue is not reproducible. The sample, however, was the simplest/minimal way for me to illustrate the problem.

Bottom line is the same code behaves differently in wxGTK port, which is unfortunate in my opinion. To avoid the difference, the code could be adapted to avoid the usage of stock items where labels are defined in client code. Instead, stock labels would be used only when the control has a standard id, but no explicitly specified label. If you agree this is reasonable, I could come up with a PR.

**Comment by vadz** [comment]
> To avoid the difference, the code could be adapted to avoid the usage of stock items where labels are defined in client code. Instead, stock labels would be used only when the control has a standard id, but no explicitly specified label. If you agree this is reasonable, I could come up with a PR.

Yes, I'd definitely expect it to work like this, i.e.

```cpp
new wxButton(this, wxID_APPLY); // Create a button with the standard label.
new wxButton(this, wxID_APPLY, "Do it!"); // Use the provided label.
```

But the problem is that it already does behave like this for me with current wxGTK, so I'm still not sure what the problem is. Could you please provide a patch to the sample showing the problem?

**Comment by artur-so** [comment]
I believe you cannot reproduce because your use case does not qualify for stock items usage:
See the condition below from `wxButton::SetLabel()` https://github.com/wxWidgets/wxWidgets/blob/a6f399a67a5c564db4f6355cb009c55eaaa8cccc/src/gtk/button.cpp#L253

(unless you use gtk 4) You should be able to reproduce with the code below or with the diff for menu I provided above:
```c++
new wxButton(this, wxID_APPLY, "&Apply"); // Use the provided label.
```
The example may seem to make little sense, but if a different than the stock one translation is needed, it could make sense in some languages (e.g. additionally specify Latin key mnemonic in Chinese).

**Comment by vadz** [comment]
Sorry, I realized that I had never answered this, and I'm not sure why this was closed.

Could you please make a patch to the internat sample showing the problem with the buttons and reopen this? The original patch still doesn't make sense to me because it doesn't use the specified context...
